### DISCUSSION TITLE
Front-facing runtime API

### DISCUSSION THREAD

**Discussion by magne4000** [magne4000/universal-middleware#76] [General]
There is a third undocumented `runtime` parameter in Middlewares and Handlers, that contains properties native to the original adapter (like `res` and `req` for node, `env` for cloudflare, etc.).

Is this the best way to represent this information and make it available to users?

**Comment by brillout** [comment]
How about this?
 - `runtime.express.{req,res}`
 - `runtime.hono.var`
 - `runtime.cloudflare.env`

Crystal clear and guaranteed no clash between runtimes.

**Reply by magne4000** [reply]
I like it, but I also would like to have some "standard" runtime keys, like `req`, `res` and `params` (for URL path segment extraction), which are already computed today.
I think that both can be added (e.g. `runtime.res` and `runtime.express.res`)

**Reply by brillout** [reply]
> I think that both can be added (e.g. `runtime.res` and `runtime.express.res`)

:+1: Sounds great, similar to Vike's plan https://github.com/vikejs/vike/issues/1268.

**Reply by brillout** [reply]
Let me know if there's anything else you'd like feedback on.

**Reply by magne4000** [reply]
There is currently a `runtime.adapter` property which contains an identifier of which adapter is currently handling the request.

For Cloudflare, it can either be `cloudflare-pages` or `cloudflare-worker`.
But I think that for `runtime`, it still makes sense to just have `runtime.cloudflare` and not `runtime['cloudflare-pages']` and `runtime['cloudflare-worker']`.

There is also a `runtime.runtime` property, which contains values like `deno`, `node`, `bun`, `workerd`, etc. which effectively are runtimes (see https://runtime-keys.proposal.wintercg.org/).

So, a request running on Cloudflare Pages would have:
```js
{
  adapter: 'cloudflare-pages',
  runtime: 'workerd',
  cloudflare: { /* ... */ }
}
```

This seems a bit too complex.

**Reply by magne4000** [reply]
All current definitions are here: https://github.com/magne4000/universal-middleware/blob/28332e3e2bc3c2730191655ae77f56ab6a33d771/packages/core/src/types.ts

**Reply by magne4000** [reply]
> This seems a bit too complex.

What bothers me the most is the fact that we cannot `runtime[runtime.adapter]` here.
But I cannot just have `runtime.adapter = 'cloudflare'`, because we need the distinction at several places.
Adding yet another property like `runtime.key = 'cloudflare'` could do the trick, but I still have the feeling that it could be simpler.

**Reply by brillout** [reply]
Maybe this?

```js
{
  adapter: 'cloudflare-pages',
  runtime: 'workerd',
  platform: 'cloudflare',
  cloudflare: { /* ... */ }
}
```

**Reply by brillout** [reply]
> Adding yet another property like `runtime.key = 'cloudflare'` could do the trick, but I still have the feeling that it could be simpler.

Some of the information is redundant (as it can be inferred) but I feel like it's okay. I guess repeating ourselves is worth it: I'd argue it's more like docs (DRY is a goal but often impractical) than internal code (that should be DRY as much as possible).

**Reply by magne4000** [reply]
I realized that the distinction between `cloudflare-pages` and `cloudflare-worker` is also necessary on the property itself (`{ "cloudflare-pages": ... }` instead of just `{ cloudflare: ... }`), because the typings are not the same.
So we can keep things simpler for now, and `runtime[runtime.adapter]` will be valid.

**Reply by brillout** [reply]
Maybe this then:

```js
{
  adapter: 'cloudflarePages',
  runtime: 'workerd',
  cloudflarePages: { /* ... */ }
}
```

Although `runtime['cloudflare-pages']` is a bit unconventional I guess it's okay.
